File: junit-tests/src/main/com/espresso/testing/EspressoTestFramework.java
package com.espresso.testing;

import java.io.FileReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.LineNumberReader;
import java.io.Reader;
import java.io.StringReader;
import java.net.URL;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.ResultSetMetaData;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.client.methods.HttpDelete;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.client.methods.HttpPut;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.DefaultHttpClient;
import org.apache.http.util.EntityUtils;
import org.apache.log4j.Logger;
import org.junit.Assert;
import org.mozilla.javascript.NativeObject;
import org.mozilla.javascript.Scriptable;
import org.mozilla.javascript.ScriptableObject;
import org.mvel2.templates.TemplateRuntime;

public class EspressoTestFramework {

	static Logger logger = Logger.getLogger(EspressoTestFramework.class);

	private final String baseUrl;
	private final String apiKey;
	private final String apiSig;
	private final String auth;

	/**
	 * 
	 * @param rootobj
	 *            - used as rootobj.getResource
	 * @param baseUrl
	 * @param apiKey
	 */
	private EspressoTestFramework(String baseUrl, String apiKey) {
		this.baseUrl = baseUrl;
		this.apiKey = apiKey;
		this.apiSig = "123";
		this.auth = "Espresso " + this.apiKey + ":" + this.apiSig;
	}

	public String toString() {
		return "URL=" + baseUrl + " apiKey=" + apiKey;
	}

	public static EspressoTestFramework create(String baseUrl, String apiKey) {
		if (!baseUrl.endsWith("/")) {
			baseUrl += "/";
			logger.debug("baseurl = " + baseUrl);
		}
		EspressoTestFramework etf = new EspressoTestFramework(baseUrl, apiKey);
		return etf;
	}

	public void logFatal(String msg) {
		logger.fatal(msg);
	}

	public void logError(String msg) {
		logger.error(msg);
	}

	public void logWarn(String msg) {
		logger.debug(msg);
	}

	public void logInfo(String msg) {
		logger.info(msg);
	}

	public void logDebug(String msg) {
		logger.debug(msg);
	}

	public void logTrace(String msg) {
		logger.trace(msg);
	}

	/**
	 * Uses MVEL2 templating, with a map derived from the enumerable properties of the specified varName object
	 * 
	 * @param varName
	 *            the javascript name to use for obtaining values - typically 'config'
	 * @param template
	 *            the string to be templated, containing template values such as @{config.foo}
	 * @return the result after templating
	 */
	public String evalTemplate(String varName, String template) {

		// this is kind of odd, but not sure how else to do this
		Scriptable scope = EspressoTestBase.scope;

		NativeObject config = (NativeObject)scope.get(varName, scope);
		Object[] ids = config.getIds();
		HashMap<String, HashMap<String, String>> templateMap = new HashMap<String, HashMap<String, String>>();
		HashMap<String, String> configMap = new HashMap<String, String>();
		templateMap.put("config", configMap);
		for (Object o : ids) {
			String propName = (String)o;
			Object val = ScriptableObject.getProperty(config, propName);
			if (null != val) {
				String javaString = val.toString();
				configMap.put(propName, javaString);
			}
		}

		Object result = TemplateRuntime.eval(template, templateMap);
		return (String)result;
	}

	public String getBaseUrl() {
		return this.baseUrl;
	}

	public String doGet(String endPoint) throws Exception {
		DefaultHttpClient client = new DefaultHttpClient();
		String url = this.baseUrl + endPoint;
		logger.trace("GET: " + url);
		HttpGet get = new HttpGet(url);
		get.setHeader("Authorization", this.auth);
		HttpResponse resp = client.execute(get);
		HttpEntity entity = resp.getEntity();
		logger.trace("Status: " + resp.getStatusLine());
		String jsonString = EntityUtils.toString(entity);
		return jsonString;
	}

	public String doPost(String endPoint, String payload) throws Exception {
		DefaultHttpClient client = new DefaultHttpClient();
		String url = this.baseUrl + endPoint;
		logger.trace("POST: " + url);
		HttpPost post = new HttpPost(url);
		post.setHeader("Authorization", this.auth);
		post.setEntity(new StringEntity(payload));
		post.setHeader("Content-Type", "application/json");
		HttpResponse resp = client.execute(post);
		logger.trace("Status: " + resp.getStatusLine());
		HttpEntity entity = resp.getEntity();
		String jsonString = EntityUtils.toString(entity);
		return jsonString;
	}

	public String doPut(String endPoint, String payload) throws Exception {
		DefaultHttpClient client = new DefaultHttpClient();
		String url = this.baseUrl + endPoint;
		logger.trace("PUT: " + url);
		HttpPut put = new HttpPut(url);
		put.setHeader("Authorization", this.auth);
		put.setEntity(new StringEntity(payload));
		put.setHeader("Content-Type", "application/json");
		HttpResponse resp = client.execute(put);
		logger.trace("Status: " + resp.getStatusLine());
		HttpEntity entity = resp.getEntity();
		String jsonString = EntityUtils.toString(entity);
		return jsonString;
	}

	public String doDelete(String endPoint) throws Exception {
		DefaultHttpClient client = new DefaultHttpClient();
		String url = this.baseUrl + endPoint;
		logger.trace("DELETE: " + url);
		HttpDelete del = new HttpDelete(url);
		del.setHeader("Authorization", this.auth);
		HttpResponse resp = client.execute(del);
		logger.trace("Status: " + resp.getStatusLine());
		HttpEntity entity = resp.getEntity();
		String jsonString = EntityUtils.toString(entity);
		return jsonString;
	}

	public void throwException(String message) {
		throw new RuntimeException(message);
	}

	public void assertion(boolean truth, CharSequence errorMessage) {
		assertion(truth, errorMessage, null);
	}

	public void assertion(boolean truth, CharSequence errorMessage, CharSequence optionalDetail) {
		String msg = errorMessage.toString();
		if (null != optionalDetail && 0 != optionalDetail.length()) {
			msg += "\n" + optionalDetail.toString();
		}
		Assert.assertTrue(msg, truth);
	}

	private class SqlStatementHolder {

		public int firstLinenumber;
		public String sqlStatement;
	}

	/**
	 * NOTE: This is a naive implementation for now.
	 * 
	 * Split given SQL into separate statements.
	 * This is actually a hard problem. Various comment types, various string quotations etc.
	 * 
	 * Oracle: For example Oracle allows a single line
	 * '/' - on a single line (unless just after another '/' in which case it is a re-execute'
	 * ';' - at end of statement (unless inside stored proc, trigger)
	 * 
	 * MySQL:
	 * ';' and end of statement;
	 * 
	 * SqlServer:
	 * GO (in various cases) on a single line
	 * 
	 * @param sql
	 * @param delim
	 * @return
	 */
	private List<SqlStatementHolder> splitSql(String sql, String delimiter, boolean isFullLineDelimiter) {
		List<SqlStatementHolder> result = new ArrayList<SqlStatementHolder>();

		StringBuffer command = null;
		LineNumberReader lineReader = new LineNumberReader(new StringReader(sql));
		String line = null;
		SqlStatementHolder holder = null;
		try {
			while ((line = lineReader.readLine()) != null) {
				if (command == null) {
					command = new StringBuffer();
					holder = new SqlStatementHolder();
					holder.firstLinenumber = lineReader.getLineNumber();
				}
				String trimmedLine = line.trim();

				int commentIdx = trimmedLine.indexOf("--"); // If the SQL ends with a comment, strip it out
				if (commentIdx >= 0) {
					trimmedLine = trimmedLine.substring(0, commentIdx);
				}
				else if (trimmedLine.length() < 1
					|| trimmedLine.startsWith("//")) {
					// Do nothing
				}
				else if (!isFullLineDelimiter && trimmedLine.endsWith(delimiter)
					|| isFullLineDelimiter && trimmedLine.equals(delimiter)) {
					command.append(trimmedLine.substring(0, trimmedLine.lastIndexOf(delimiter)));
					holder.sqlStatement = command.toString();
					result.add(holder);
					command = null;
					holder = null;
				}
				else {
					command.append(trimmedLine);
					command.append('\n');
				}
			}
		}
		catch (IOException e) {
			return null;
		}

		return result;
	}

	/**
	 * 
	 * @param url
	 * @param username
	 * @param password
	 * @param sql
	 *            a single sql select statement
	 * @param ignoreErrors
	 */
	public String[][] executeQuery(String url, String username, String password, String sql, boolean ignoreErrors) {
		Connection conn = null;
		Statement stmt = null;
		try {
			conn = DriverManager.getConnection(url, username, password);
			stmt = conn.createStatement();
			ResultSet rs = stmt.executeQuery(sql);
			ArrayList<String[]> result = new ArrayList<String[]>();
			ResultSetMetaData rsmeta = rs.getMetaData();
			int numCols = rsmeta.getColumnCount();
			String[] columnNames = new String[numCols];
			String[] columnTypes = new String[numCols];
			for (int i = 0; i < numCols; ++i) {
				columnNames[i] = rsmeta.getColumnName(1 + i);
				columnTypes[i] = rsmeta.getColumnTypeName(1 + i);
			}

			result.add(columnNames);

			result.add(columnTypes);
			while (rs.next()) {
				String[] vals = new String[numCols];
				for (int i = 0; i < numCols; ++i) {
					vals[i] = rs.getString(1 + i);
				}
				result.add(vals);
			}

			String[][] newResult = result.toArray(new String[0][0]);
			return newResult;
		}
		catch (SQLException e) {
			logger.error("Database " + url, e);
			throwException(e.getMessage());
			return null;
		}
		finally {
			if (null != stmt) {
				try {
					stmt.close();
				}
				catch (Exception ex) {
				}
				stmt = null;
			}
			if (null != conn) {
				try {
					conn.close();
				}
				catch (Exception ex) {
				}
				conn = null;
			}
		}
	}

	public void executeSql(String url, String username, String password, String sql, boolean ignoreErrors) {
		Connection conn = null;
		Statement stmt = null;

		try {
			List<SqlStatementHolder> sqlStatements = this.splitSql(sql, ";", false);
			conn = DriverManager.getConnection(url, username, password);
			stmt = conn.createStatement();
			for (SqlStatementHolder holder : sqlStatements) {
				try {
					logger.trace("SQL: " + holder.sqlStatement);
					stmt.execute(holder.sqlStatement);
				}
				catch (SQLException ex) {
					String msg = "Error near line " + holder.firstLinenumber + "\n" + holder.sqlStatement + "\n" + ex.getMessage();
					if (ignoreErrors) {
						logger.error("SQL IGNORE: " + msg);
					}
					else {
						logger.error("SQL: " + msg);
						throwException(msg);
					}
				}
			}
		}
		catch (SQLException e) {
			logger.error("Database " + url, e);
			throwException(e.getMessage());
		}
		finally {
			if (null != stmt) {
				try {
					stmt.close();
				}
				catch (Exception e) {
				}
			}
			stmt = null;
			if (null != conn) {
				try {
					conn.close();
				}
				catch (Exception e) {
				}
			}
			conn = null;
		}
	}

	public static String readFile(String path) {
		logger.debug("READ: " + path);
		Reader rdr = null;
		try {
			rdr = new FileReader(path);

			StringBuilder sb = new StringBuilder();
			int c;
			while (-1 != (c = rdr.read())) {
				sb.append((char)c);
			}
			rdr.close();
			return sb.toString();
		}
		catch (IOException e) {
			if (null != rdr) {
				try {
					rdr.close();
				}
				catch (Exception e2) {
				}
			}
			e.printStackTrace();
			throw new RuntimeException(e);
		}
	}

	public static String readStream(String tag, InputStream stream) {
		logger.debug("READSTREAM: " + tag);
		Reader rdr = null;
		try {
			rdr = new InputStreamReader(stream);

			StringBuilder sb = new StringBuilder();
			int c;
			while (-1 != (c = rdr.read())) {
				sb.append((char)c);
			}
			rdr.close();
			return sb.toString();
		}
		catch (IOException e) {
			if (null != rdr) {
				try {
					rdr.close();
				}
				catch (Exception e2) {
				}
			}
			e.printStackTrace();
			throw new RuntimeException(e);
		}
	}

	public static String readResource(String javascriptResourceName) {
		try {
			String filename = javascriptResourceName;
			InputStream is = null;
			URL url = null;
			if (null != EspressoTestBase.rootClass) {
				url = EspressoTestBase.rootClass.getResource(javascriptResourceName);
				is = EspressoTestBase.rootClass.getResourceAsStream(javascriptResourceName);
			}
			else {
				url = ClassLoader.getSystemClassLoader().getResource(javascriptResourceName);
				is = ClassLoader.getSystemClassLoader().getResourceAsStream(javascriptResourceName);
			}

			if (null != url) {
				filename = url.getPath();
			}

			if (null != is) {
				String js = EspressoTestFramework.readStream(filename, is);
				return js;
			}
			return null;
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}
}

File: junit-tests/src/main/com/espresso/testing/EspressoTestFramework.js
var defaultEnv = "dblocal";
var defaultSecondary = "localhost";

function espressoTestFramework(baseUrl, apiKey) {
	"use strict";

	return (function(myBaseUrl, myApiKey) {

		var etf = Packages.com.espresso.testing.EspressoTestFramework.create(myBaseUrl, myApiKey);

		var self = {};

		self.readResource = function(resourceName) {
			var resource = Packages.com.espresso.testing.EspressoTestFramework.readResource(resourceName);
			return resource;
		};

		self.evalTemplate = function(varName, template) {
			print("before temp: " + template);
			var result = etf.evalTemplate(varName.toString(), template.toString());
			print("templated: " + result);
			return result;
		};

		self.getBaseUrl = function() {
			return "" + etf.baseUrl;
		};

		self.doGet = function(endPoint) {
			etf.logTrace("Get " + endPoint);
			var json = etf.doGet(endPoint);
			return JSON.parse(json);
		};

		self.doPost = function(endPoint, jsonPayload) {
			etf.logTrace("Post " + endPoint);
			var json = etf.doPost(endPoint, JSON.stringify(jsonPayload));
			try {
				return JSON.parse(json);
			} catch (e) {
				etf.logError("Error parsing " + e + "\n" + jsonPayload);
				throw e;
			}
		};

		self.insert = function(endPoint, jsonPayload) {
			var result = self.doPost(endPoint, jsonPayload);
			var stringResult = JSON.stringify(result, null, 2);
			etf.logWarn("inserted " + stringResult);
			etf.assertion(result && result.hasOwnProperty("statusCode"), "Expected statusCode property in result", stringResult);
			etf.assertion(201 == result.statusCode, "Expected 201 status code - got " + result.statusCode + " while inserting " + endPoint, stringResult);
			return result;
		};

		self.doPut = function(endPoint, jsonPayload) {
			etf.logTrace("Put " + endPoint);
			var json = etf.doPut(endPoint, JSON.stringify(jsonPayload));
			return JSON.parse(json);
		};

		self.doDelete = function(endPoint) {
			etf.logTrace("Delete " + endPoint);
			var json = etf.doDelete(endPoint);
			return JSON.parse(json);
		};

		self.assertion = function(truth, failMessage, optionalObj) {
			if (!failMessage)
				failMessage = "";
			var optionalString;
			if (optionalObj) {
				if ("object" === typeof optionalObj) {
					optionalString = JSON.stringify(optionalObj, null, 2);
				} else {
					optionalString = optionalObj.toString();
				}
			} else {
				optionalString = null;
			}

			etf.assertion(!!truth, failMessage, optionalString);
		};

		self.logFatal = function(msg) {
			etf.logFatal(msg);
		};

		self.logError = function(msg) {
			etf.logError(msg);
		};

		self.logWarn = function(msg) {
			etf.logWarn(msg);
		};

		self.logInfo = function(msg) {
			etf.logInfo(msg);
		};

		self.logDebug = function(msg) {
			etf.logDebug(msg);
		};

		self.logTrace = function(msg) {
			etf.logTrace(msg);
		};

		/*
		 * Finds by name any and all that were inserted
		 */
		self.findInserted = function(txsummary, resourceName) {
			var result = new Array();
			var idx = undefined;
			for (idx in txsummary) {
				var item = txsummary[idx];
				var metadata = item["@metadata"];
				if (resourceName === metadata.resource && "INSERT" === metadata.verb) {
					result.push(item);
				}
			}
			return result;
		};

		/*
		 * Finds by name any and all that were updated
		 */
		self.findUpdated = function(txsummary, resourceName) {
			var result = new Array();
			var idx = undefined;
			for (idx in txsummary) {
				var item = txsummary[idx];
				var metadata = item["@metadata"];
				if (resourceName === metadata.resource && "UPDATE" === metadata.verb) {
					result.push(item);
				}
			}
			return result;
		};

		self.executeQuery = function(url, username, password, sql) {
			// TODO - this is not complete - the java currently returns strings,
			// this will need to be
			// enhanced.
			var objs = etf.executeQuery(url, username, password, sql, false);
			var result = {
				names : [],
				types : [],
				data : []
			};

			// TODO - this is NOT right - need to morph to better javascript
			// types
			// NOT just strings.
			var val = undefined;
			for ( var i in objs[0]) {
				val = "" + objs[0][i];
				result.names[i] = val;
			}
			for ( var i in objs[1]) {
				val = "" + objs[1][i];
				result.types[i] = val;
			}
			for ( var i = 2; i < objs.length; ++i) {
				result.data[i - 2] = [];
				for ( var j in objs[i]) {
					var val = "" + objs[i][j];
					result.data[i - 2][j] = val;
				}
			}
			return result;
		};

		self.executeSql = function(url, username, password, sql) {
			try {
				etf.executeSql(url, username, password, sql, false);
			} catch (e) {
				print("error: " + e);
				print("url: " + url);
				print("username: " + username);
				print("password: " + password);
				print("sql: " + password);
				self.logError(e);
				throw e;
			}
		};

		self.executeSqlIgnoreErrors = function(url, username, password, sql) {
			print("dmh url: " + url);
			print("dmh username: " + username);
			print("dmh password: " + password);
			print("dmh sql: " + sql);
			etf.executeSql(url, username, password, sql, true);
		};
		return self;
	})(baseUrl, apiKey);
}

// read the environment and java system properties
// ensure that env.KahunaEnvironment and env.KahunaEnvironmentSecondary is set
// with a value
// first from system environment
// failing that from java system properties
// failing that a default value of dblocal and localhost
(function(global) {
	"use strict";
	function convertJavaMap(map) {
		var result = {};
		var keySet = map.keySet();
		var keySetIter = keySet.iterator();
		while (keySetIter.hasNext()) {
			var s = "" + keySetIter.next();
			var v = "" + map.get(s);
			result[s] = v;
		}
		return result;
	}

	var newenv = convertJavaMap(java.lang.System.getenv());
	var newSystemProperties = convertJavaMap(java.lang.System.getProperties());

	if ('undefined' === typeof newenv.KahunaEnvironment) {
		if ('undefined' === typeof newSystemProperties.KahunaEnvironment) {
			newenv.KahunaEnvironment = defaultEnv;
			print('Using default KahunaEnvironment of ' + newenv.KahunaEnvironment);
		} else {
			print("Using java system properties for environment " + newSystemProperties.KahunaEnvironment);
			newenv.KahunaEnvironment = newSystemProperties.KahunaEnvironment;
		}
	}

	if ('undefined' === typeof newenv.KahunaEnvironmentSecondary) {
		if ('undefined' === typeof newSystemProperties.KahunaEnvironmentSecondary) {
			newenv.KahunaEnvironmentSecondary = defaultSecondary;
			print('Using default KahunaEnvironmentSecondary of ' + newenv.KahunaEnvironmentSecondary);
		} else {
			print("Using java system properties for environment secondary" + newSystemProperties.KahunaEnvironmentSecondary);
			newenv.KahunaEnvironmentSecondary = newSystemProperties.KahunaEnvironmentSecondary;
		}
	}

	global.env = newenv;
	global.systemProperties = newSystemProperties;
})(this);

function loadConfig(global, base) {
	"use strict";
	// typically loadConfig(this, "config")
	// and loadConfig(this, "/config");
	var primary = env.KahunaEnvironment;
	var secondary = env.KahunaEnvironmentSecondary;

	print("Environment: " + primary + " / " + secondary);

	var readResource = function(resourceName) {
		var str = Packages.com.espresso.testing.EspressoTestFramework.readResource(resourceName);
		if (undefined == typeof str || !str) {
			print("Not found: " + resourceName);
			str = null;
		}
		return str;
	};

	if ("undefined" === typeof global.config || !global.config) {
		print("Creating config");
		global.config = {};
	}

	var tmp = null;
	var filename = base + "/" + primary + ".json";
	var primaryConfig = readResource(filename);
	if (primaryConfig) {
		print("Parsing: " + filename);
		primaryConfig = JSON.parse(JSON.minify(primaryConfig));
		for (tmp in primaryConfig) {
			global.config[tmp] = primaryConfig[tmp];
		}
	}

	filename = base + "/" + primary + "-" + secondary + ".json";
	var secondaryConfig = readResource(filename);
	if (secondaryConfig) {
		print("Parsing: " + filename);
		secondaryConfig = JSON.parse(JSON.minify(secondaryConfig));

		for (tmp in secondaryConfig) {
			global.config[tmp] = secondaryConfig[tmp];
		}
	}
}

// load the standard common config into the global object
loadConfig(this, "/config");
